### PR TITLE
Don't evaluate regexes twice

### PR DESCRIPTION
**Pull request by iamdanfox**
## Before this PR

The GC problems on the service I work on are a lot better, but not completely solved. There are some sensible domain-related changes we could make, but something that stuck out to me when looking at the JFRs was how much costly it was to just parse version numbers (in terms of CPU time and also allocation)

image

Interestingly we've had this comment `Note: Can avoid duplicate regex evaluate if performance ever becomes a concern.`

## After this PR
==COMMIT_MSG==
Parsing an `OrderableSlsVersion` no longer runs the regex matcher twice
==COMMIT_MSG==

```
BEFORE


Benchmark                                                         (versionString)  Mode  Cnt     Score      Error   Units
SlsVersionBenchmark.safeValueOf                                           RELEASE  avgt    3   278.398 ±    1.942   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                            RELEASE  avgt    3  7413.325 ±   80.262  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                       RELEASE  avgt    3   632.032 ±    0.011    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                   RELEASE  avgt    3  7448.247 ±  240.664  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm              RELEASE  avgt    3   635.010 ±   15.795    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space               RELEASE  avgt    3     0.033 ±    0.043  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm          RELEASE  avgt    3     0.003 ±    0.004    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                 RELEASE  avgt    3   177.000             counts
SlsVersionBenchmark.safeValueOf:·gc.time                                  RELEASE  avgt    3    97.000                 ms
SlsVersionBenchmark.safeValueOf                                          SNAPSHOT  avgt    3   909.686 ±  177.282   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                           SNAPSHOT  avgt    3  4797.870 ±  932.620  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                      SNAPSHOT  avgt    3  1336.090 ±    0.040    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                  SNAPSHOT  avgt    3  4821.985 ±   58.242  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm             SNAPSHOT  avgt    3  1342.908 ±  263.204    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space              SNAPSHOT  avgt    3     0.023 ±    0.048  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm         SNAPSHOT  avgt    3     0.006 ±    0.012    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                SNAPSHOT  avgt    3   138.000             counts
SlsVersionBenchmark.safeValueOf:·gc.time                                 SNAPSHOT  avgt    3    75.000                 ms
SlsVersionBenchmark.safeValueOf                                                RC  avgt    3   534.555 ±  135.397   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                                 RC  avgt    3  5473.985 ± 1400.198  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                            RC  avgt    3   896.061 ±    0.016    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                        RC  avgt    3  5520.608 ± 1141.179  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm                   RC  avgt    3   903.744 ±  168.509    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space                    RC  avgt    3     0.031 ±    0.114  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm               RC  avgt    3     0.005 ±    0.019    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                      RC  avgt    3   158.000             counts
SlsVersionBenchmark.safeValueOf:·gc.time                                       RC  avgt    3    85.000                 ms
SlsVersionBenchmark.safeValueOf                                       RC_SNAPSHOT  avgt    3   678.839 ±  170.689   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                        RC_SNAPSHOT  avgt    3  4197.744 ± 1011.516  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                   RC_SNAPSHOT  avgt    3   872.060 ±    0.017    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space               RC_SNAPSHOT  avgt    3  4232.266 ± 1097.996  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm          RC_SNAPSHOT  avgt    3   879.280 ±  219.905    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space           RC_SNAPSHOT  avgt    3     0.020 ±    0.042  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm      RC_SNAPSHOT  avgt    3     0.004 ±    0.008    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                             RC_SNAPSHOT  avgt    3   121.000             counts
SlsVersionBenchmark.safeValueOf:·gc.time                              RC_SNAPSHOT  avgt    3    65.000                 ms
SlsVersionBenchmark.safeValueOf                                             DIRTY  avgt    3   632.662 ±  956.822   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                              DIRTY  avgt    3  4522.205 ± 6689.555  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                         DIRTY  avgt    3   872.060 ±    0.009    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                     DIRTY  avgt    3  4545.028 ± 7768.420  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm                DIRTY  avgt    3   875.854 ±  209.024    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space                 DIRTY  avgt    3     0.019 ±    0.056  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm            DIRTY  avgt    3     0.004 ±    0.005    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                   DIRTY  avgt    3   130.000             counts
SlsVersionBenchmark.safeValueOf:·gc.time                                    DIRTY  avgt    3    73.000                 ms

Process finished with exit code 0

AFTER


Benchmark                                                         (versionString)  Mode  Cnt     Score       Error   Units
SlsVersionBenchmark.safeValueOf                                           RELEASE  avgt    3   194.011 ±   501.334   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                            RELEASE  avgt    3  7094.521 ± 17288.513  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                       RELEASE  avgt    3   416.019 ±     0.007    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                   RELEASE  avgt    3  7120.312 ± 16316.958  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm              RELEASE  avgt    3   417.846 ±    82.823    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space               RELEASE  avgt    3     0.024 ±     0.136  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm          RELEASE  avgt    3     0.001 ±     0.005    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                 RELEASE  avgt    3   140.000              counts
SlsVersionBenchmark.safeValueOf:·gc.time                                  RELEASE  avgt    3    93.000                  ms
SlsVersionBenchmark.safeValueOf                                          SNAPSHOT  avgt    3   714.961 ±   402.728   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                           SNAPSHOT  avgt    3  5119.310 ±  2900.577  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                      SNAPSHOT  avgt    3  1120.075 ±     0.029    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                  SNAPSHOT  avgt    3  5139.262 ±  1851.504  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm             SNAPSHOT  avgt    3  1124.726 ±   287.525    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space              SNAPSHOT  avgt    3     0.023 ±     0.073  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm         SNAPSHOT  avgt    3     0.005 ±     0.013    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                SNAPSHOT  avgt    3   147.000              counts
SlsVersionBenchmark.safeValueOf:·gc.time                                 SNAPSHOT  avgt    3    81.000                  ms
SlsVersionBenchmark.safeValueOf                                                RC  avgt    3   443.385 ±   577.418   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                                 RC  avgt    3  5024.689 ±  6824.128  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                            RC  avgt    3   680.038 ±     0.004    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                        RC  avgt    3  5057.595 ±  7982.883  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm                   RC  avgt    3   684.097 ±   145.258    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space                    RC  avgt    3     0.016 ±     0.110  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm               RC  avgt    3     0.002 ±     0.012    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                      RC  avgt    3   120.000              counts
SlsVersionBenchmark.safeValueOf:·gc.time                                       RC  avgt    3    77.000                  ms
SlsVersionBenchmark.safeValueOf                                       RC_SNAPSHOT  avgt    3   711.226 ±   410.655   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                        RC_SNAPSHOT  avgt    3  4010.460 ±  2271.405  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                   RC_SNAPSHOT  avgt    3   872.060 ±     0.036    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space               RC_SNAPSHOT  avgt    3  4026.378 ±  2207.875  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm          RC_SNAPSHOT  avgt    3   875.543 ±    60.157    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space           RC_SNAPSHOT  avgt    3     0.016 ±     0.072  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm      RC_SNAPSHOT  avgt    3     0.003 ±     0.014    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                             RC_SNAPSHOT  avgt    3   115.000              counts
SlsVersionBenchmark.safeValueOf:·gc.time                              RC_SNAPSHOT  avgt    3    66.000                  ms
SlsVersionBenchmark.safeValueOf                                             DIRTY  avgt    3   616.208 ±   740.972   ns/op
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate                              DIRTY  avgt    3  4634.911 ±  5512.971  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.alloc.rate.norm                         DIRTY  avgt    3   872.055 ±     0.111    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space                     DIRTY  avgt    3  4645.809 ±  5495.985  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Eden_Space.norm                DIRTY  avgt    3   874.135 ±    88.477    B/op
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space                 DIRTY  avgt    3     0.019 ±     0.087  MB/sec
SlsVersionBenchmark.safeValueOf:·gc.churn.G1_Survivor_Space.norm            DIRTY  avgt    3     0.004 ±     0.014    B/op
SlsVersionBenchmark.safeValueOf:·gc.count                                   DIRTY  avgt    3   122.000              counts
SlsVersionBenchmark.safeValueOf:·gc.time                                    DIRTY  avgt    3    72.000                  ms

```

## Possible downsides?
- this might be excessive... my normal rule of thumb is to avoid microoptimising unless you can get at least an order of magnitude

### CODE REVIEW

File: sls-versions/src/main/java/com/palantir/sls/versions/SlsVersionMatcherParser.java [hunk 1]
```diff
@@ -121,7 +121,7 @@ private static long number(String string, int startIndex) {
             try {
                 return ok(next, Integer.parseUnsignedInt(string.substring(startIndex, next)));
             } catch (NumberFormatException e) {
-                if (e.getMessage().endsWith("exceeds range of unsigned int.")) {
+                if (e.getMessage() != null && e.getMessage().endsWith("exceeds range of unsigned int.")) {
```
Comment (carterkozak) [inline sls-versions/src/main/java/com/palantir/sls/versions/SlsVersionMatcherParser.java]: Might as well simplify things for the compiler:
```suggestion
                String message = e.getMessage();
                if (message != null && message.endsWith("exceeds range of unsigned int.")) {
```

File: sls-versions/src/main/java/com/palantir/sls/versions/RegexParser.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        this.pattern = pattern;
+    }
+
+    static RegexParser of(String regex) {
+        Pattern pattern = Pattern.compile(regex);
+        return new RegexParser(pattern);
+    }
+
+    @Nullable
+    public Matcher tryParse(String string) {
```
Comment (carterkozak) [inline sls-versions/src/main/java/com/palantir/sls/versions/RegexParser.java]: nit
```suggestion
   Matcher tryParse(String string) {
```

**Comment by carterkozak** [comment]
Going to give this a proper read through tomorrow

File: build.gradle [hunk 3]
```diff
@@ -43,5 +45,20 @@ subprojects {
...
+    }
+
+    plugins.withId('com.palantir.baseline-error-prone', {
+        dependencies {
+            compileOnly 'com.google.code.findbugs:jsr305'
+            annotationProcessor 'com.uber.nullaway:nullaway'
+        }
+
+        tasks.withType(JavaCompile) {
+            options.errorprone.errorproneArgs.add '-XepOpt:NullAway:AnnotatedPackages=com.palantir'
```
Comment (carterkozak) [inline build.gradle:60]: This should probably be either a separate plugin, or config option in the errorprone extension. Not suggesting we change this, just musing on potential improvements.

Comment (iamdanfox) [reply build.gradle:60]: Yeah agreed - I think @schlosna had the same feeling when he had to add lots of boilerplate to tritium to opt-in to extra strictness

File: sls-versions/src/main/java/com/palantir/sls/versions/RegexParser.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        this.pattern = pattern;
+    }
+
+    static RegexParser of(String regex) {
+        Pattern pattern = Pattern.compile(regex);
+        return new RegexParser(pattern);
+    }
+
+    @Nullable
+    Matcher tryParse(String string) {
```
Comment (carterkozak) [inline sls-versions/src/main/java/com/palantir/sls/versions/RegexParser.java]: Let's javadoc this explaining that a matcher is only returned if it matches.

File: sls-versions/src/main/java/com/palantir/sls/versions/OrderableSlsVersion.java [hunk 5]
```diff
@@ -54,37 +53,35 @@ public static OrderableSlsVersion valueOf(String value) {
...
-        SlsVersionType type = getTypeOrNull(value);
-        if (type == null) {
+        SlsVersionType finalType = ORDERED_VERSION_TYPES[0];
+        Matcher groups = null;
+        for (SlsVersionType type : ORDERED_VERSION_TYPES) {
+            Matcher maybeGroups = type.getParser().tryParse(value);
+            if (maybeGroups != null) {
+                finalType = type;
+                groups = maybeGroups;
+                break;
```
Comment (carterkozak) [inline sls-versions/src/main/java/com/palantir/sls/versions/OrderableSlsVersion.java]: Might be cleaner to split method in half, and replace these 3 lines with `return callSecondHalf(type, maybeGroups);`

**Comment by carterkozak** [comment]
lgtm pending splitting the parse method

File: sls-versions/src/main/java/com/palantir/sls/versions/OrderableSlsVersion.java (hunk 5 above)
Comment (iamdanfox) [reply sls-versions/src/main/java/com/palantir/sls/versions/OrderableSlsVersion.java]: great point - reads much nicer

**Review by carterkozak — APPROVED** [review]
thanks!

File: build.gradle (hunk 3 above)
Comment (schlosna) [reply build.gradle:60]: Yeah, having a simple NullAway opt-in inside gradle-baseline would be nice. It is definitely much easier to add NullAway to projects early on than to retrofit after the fact, and does have [some false positives around mock injections for unit tests and JMH](https://github.com/palantir/tritium/search?l=Java&q=NullAway).

**Pull request merged**